test/cql-pytest/test_group_by.py

```diff
@@ -38,13 +38,47 @@ def test_group_by_partition(cql, table1):
 
 # Adding a LIMIT should be honored.
 # Reproduces #17237 - more results than the limit were generated.
-@pytest.mark.xfail(reason="issue #17237")
 def test_group_by_partition_with_limit(cql, table1):
     # "LIMIT 1" should return only one of the two matches (0,0,0,1) and
     # (1,0,0,5). The partition key 1 happens to be the first one in
     # murmur3 order, so this is what should be returned.
     assert {(1,0,0,5)} == set(cql.execute(f'SELECT p,c1,c2,v FROM {table1} GROUP BY p LIMIT 1'))
 
+def test_group_by_partition_count_with_limit(cql, test_keyspace):
+    with new_test_table(cql, test_keyspace, "a int, b int, c int, PRIMARY KEY (a, b)") as table:
+        # Insert 4 rows with the same partition key, but different clustering keys
+        # 29, 31, 37, 41 are primes, so they have no common factors with each
+        # other, and will not randomly become a result of multiplication.
+        for i in range(29):
+            cql.execute(f"insert into {table} (a,b,c) values ({29},{i+1},29);");
+        for i in range(31):
+            cql.execute(f"insert into {table} (a,b,c) values ({31},{i+1},31);");
+        for i in range(37):
+            cql.execute(f"insert into {table} (a,b,c) values ({37},{i+1},37);");
+        for i in range(41):
+            cql.execute(f"insert into {table} (a,b,c) values ({41},{i+1},41);");
+        query_result = list(cql.execute(f'SELECT a FROM {table} GROUP BY a'))
+        # We should get 4 rows, one for each partition key
+        assert len(query_result) == 4
+        assert {29, 31, 37, 41} == set(map(lambda x: x[0], query_result))
+        # Adding a LIMIT should be honored.
+        for i in range(1,4):
+            query_result = set(map(lambda x: x[0], cql.execute(f'SELECT a FROM {table} GROUP BY a LIMIT {i}')))
+            # We should get only {i} rows, one of the partition keys
+            assert len(query_result) == i
+            assert query_result.issubset({29, 31, 37, 41})
+        # COUNT should return proper count values (e.g. 29 for partition key 29)
+        query_result = list(cql.execute(f'SELECT COUNT(a) FROM {table} GROUP BY a'))
+        # We should get 4 rows, one for each partition key
+        assert len(query_result) == 4
+        assert {29, 31, 37, 41} == set(map(lambda x: x[0], query_result))
+        # Adding a LIMIT should be honored.
+        for i in range(1,4):
+            query_result = set(map(lambda x: x[0], cql.execute(f'SELECT COUNT(a) FROM {table} GROUP BY a LIMIT {i}')))
+            # We should get only {i} rows, one of the partition keys
+            assert len(query_result) == i
+            assert query_result.issubset({29, 31, 37, 41})
+
```

Comment:
@KrzaQ I thought you'd add a test for selection with paging; did you find an existing one?

Reply:
My concern is that the test you're referencing is a single -node test, and in your added test you install multiple nodes. It's worth having an extra test which uses paging + multiple nodes given the cost of setting up a multi-node cluster is high.

Reply:
@KrzaQ I looked at the test you referenced (a test translated from Cassandra unit tests), and although the test is called testSelectCountPaging, its actual code doesn't do any paging! It uses a secondary index, to further complicate things, but it doesn't do paging. It also refers to https://issues.apache.org/jira/browse/CASSANDRA-6579, which has nothing to do with paging.

I'll send a PR in a few minutes with additional tests, including one with paging.

@kostja, yes, all those tests are single node. In theory they can miss bugs that have to do with how reconciliation happens on a coordinator, or how a coordinator sends a request with a LIMIT to replicas. That's an interesting thing to check: E.g., what happens if two replicas have slightly different rows, and I run on a coordinator a SELECT with LIMIT 10, is it possible that each of the different nodes returns a different set of 10 results and together more than 10 results are created? Is the opposite possible - one replica returns 10 results but because of a tombstone from another replica, the final results will be fewer than 10? It's definitely worthwhile writing new tests for these things in the topology framework. But it doesn't replace the single-node tests.

P.S. The idea that "the cost of setting up a multi-node cluster is high" hurt dtest considerably, and now hurts the topology framework. We should be able to create a 3-node cluster once, and run hundreds of tests on it. Just like in the 1-node tests in cql-pytest. https://github.com/scylladb/scylladb/issues/16478 should be done to make all of this efficient.